filebeat/input/filestream/input_test.go

```diff
@@ -0,0 +1,332 @@
+	pipeline := &mockPipelineConnector{}
+
+	var wg sync.WaitGroup
+	go func() {
+		wg.Add(1)
+		defer wg.Done()
+
+		inp.Run(inputCtx, pipeline)
+	}()
+
+	// first event has made it successfully
+	waitUntilEventCount(1, pipeline)
+	// check registry
+	checkOffsetInRegistry(t, testStore, testlogPath, len([]byte("first log line\n")))
+
+	// the second log line will not be picked up as scan_interval is set to one day.
+	err = ioutil.WriteFile(testlogPath, []byte("first log line\nsecond log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
+	time.Sleep(1 * time.Second)
+
+	// only one event is read
+	waitUntilEventCount(1, pipeline)
+
+	cancelInput()
+	wg.Wait()
+
+	checkOffsetInRegistry(t, testStore, testlogPath, len([]byte("first log line\n")))
+}
+
+func checkOffsetInRegistry(t *testing.T, store loginp.StateStore, filepath string, expectedOffset int) {
+	fi, err := os.Stat(filepath)
+	if err != nil {
+		t.Fatalf("cannot stat file when cheking for offset: %+v", err)
+	}
+
+	identifier, _ := newINodeDeviceIdentifier(nil)
+	src := identifier.GetSource(loginp.FSEvent{Info: fi, Op: loginp.OpCreate, NewPath: filepath})
+	entry := getRegistryState(t, store, src.Name())
+
+	require.Equal(t, expectedOffset, entry.Cursor.Offset)
+}
+
+func getInputFromConfig(t *testing.T, s loginp.StateStore, c map[string]interface{}) input.Input {
+	manager := &loginp.InputManager{
+		Logger:     logp.L(),
+		StateStore: s,
+		Type:       pluginName,
+		Configure:  configure,
+	}
+
+	inp, err := manager.Create(common.MustNewConfigFrom(c))
+	if err != nil {
+		t.Fatalf("cannot create filestream input: %+v", err)
+	}
+	return inp
+}
+
+type registryEntry struct {
+	Cursor struct {
+		Offset int `json:"offset"`
+	} `json:"cursor"`
+}
+
+func getRegistryState(t *testing.T, s loginp.StateStore, key string) registryEntry {
+	inputStore, _ := s.Access()
+
+	var e registryEntry
+	err := inputStore.Get(key, &e)
+	if err != nil {
+		t.Fatalf("error when getting expected key '%s' from store: %+v", key, err)
+	}
+
+	return e
+}
+
+func waitUntilEventCount(count int, pipeline *mockPipelineConnector) {
+	for {
+		for _, c := range pipeline.clients {
+			if len(c.GetEvents()) == count {
+				return
+			}
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+}
+
+type testInputStore struct {
+	registry *statestore.Registry
+}
+
+func openTestStatestore(t *testing.T, tmpDir string) loginp.StateStore {
+	memlog, err := memlog.New(logp.L(), memlog.Settings{
+		Root:     tmpDir,
+		FileMode: 0644,
+	})
+	if err != nil {
+		t.Fatalf("cannot open statestore: %+v", err)
+	}
+
+	r := statestore.NewRegistry(memlog)
+	return &testInputStore{
+		registry: r,
+	}
+}
+
+func (s *testInputStore) Close() {
+	s.registry.Close()
+}
+
+func (s *testInputStore) Access() (*statestore.Store, error) {
+	return s.registry.Get("filebeat")
+}
+
+func (s *testInputStore) CleanupInterval() time.Duration {
+	return 24 * time.Hour
+}
+
+type mockClient struct {
```

Comment:
`libbeat/publisher/testing` contains some helpers. Any helper you can reuse or shall we add some?

Reply:
I have added these mocks because I wanted to include the ACK handler in the test. I am fine with adding these to `libbeat/publisher/testing`. My initial reason behind not moving the mocks there was that I did not want to generalize these too early, rather keep it self-contained in the test.